qiskit_nature/problems/second_quantization/molecular/molecular_problem.py

```diff
@@ -0,0 +1,111 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The Molecular Problem class."""
+from typing import List, Optional, Tuple
+
+import numpy as np
+
+from qiskit_nature.drivers.qmolecule import QMolecule
+from qiskit_nature.drivers import FermionicDriver
+from qiskit_nature.operators import FermionicOp
+from qiskit_nature.operators.second_quantization import SecondQuantizedOp
+from qiskit_nature.problems.second_quantization.molecular.integrals_calculators \
+    .angular_momentum_integrals_calculator import \
+    calc_total_ang_momentum_ints
+from qiskit_nature.problems.second_quantization.molecular.fermionic_op_builder import \
+    build_fermionic_op, \
+    build_ferm_op_from_ints
+from qiskit_nature.problems.second_quantization.molecular.integrals_calculators \
+    .magnetization_integrals_calculator \
+    import calc_total_magnetization_ints
+from qiskit_nature.problems.second_quantization.molecular.integrals_calculators \
+    .particle_number_integrals_calculator \
+    import \
+    calc_total_particle_num_ints
+from qiskit_nature.transformations.second_quantization import BaseTransformer
+
+
+class MolecularProblem:
+    """Molecular Problem"""
+
+    def __init__(self, fermionic_driver: FermionicDriver,
+                 second_quantized_transformations: Optional[List[BaseTransformer]]):
+        """
+
+        Args:
+            fermionic_driver: A fermionic driver encoding the molecule information.
+            second_quantized_transformations: A list of second quantized transformations to be
+            applied to the molecule.
+        """
+        self.driver = fermionic_driver
+        self.transformers = second_quantized_transformations
+        self._q_molecule = self.driver.run()
```

Comment:
I don't think you should call `driver.run()` during the construction. Rather, do so in `second_q_ops` and store the result. If the operators are queried multiple times you can re-use the stored `q_molecule`.

Reply:
If the driver can be updated, which I think can be done via Molecule class, then asking for the 2nd q op should given back something consistent with the driver. Hence calling the run and transforms etc (maybe transforms can be updated too via some setter/getter) in the method to get the 2nd q op seems like the right thing to do. Given the driver can logically be updated without this class knowing I am not sure storing/caching anything here is a good idea.

Also since the integrals are affected by the transforms storing some self._num_modes upfront is problematic,

Reply:
That is a good point! I still think that storing the `q_molecule` can have a very significant performance benefit in some cases. But we could probably invalidate any stored values based on whether the driver was updated. But let's leave that to the driver refactoring.

Reply:
You are right that there is a danger here. But how about we make driver and transformers immutable in `MolecularProblem`? This way, a user would create a new instance of this class for each `driver` and/or `transformers`. This might be a cleaner solution that allowing things to change.

Reply:
I think this may conflict with some of the plans for the PES samplers. @pbark should know more about this.

Reply:
In the overall scheme of things the amount of time it has taken so far in the classical computation with the driver is but a fraction of the overall time. 

As far as making them immutable, since you can change Molecule, which is given to a driver, making the driver itself not be able to be changed in the problem will make no difference. It looks at the Molecule when asked to run and has no hooks to know things were changed in it ahead of that point.

I would just do the simple thing and run the driver and transforms when needed when creating the 2QOp. Later performance tests can be done to look at places where things are worth improving that will make a significant difference.